org.eclipse.jdt.core/formatter/org/eclipse/jdt/internal/formatter/CommentsPreparator.java

```diff
@@ -793,6 +807,252 @@ private void alignJavadocTag(List<Token> tagTokens, int paramNameAlign, int desc
 		}
 	}
 
+	private void handleMarkdown(TagElement node) {
+		if (!(node.getParent() instanceof Javadoc javaDoc) || !javaDoc.isMarkdown()
+				|| !this.options.comment_format_markdown_comment) {
+			return;
+		}
+
+		String text = this.tm.toString(node);
+
+		Matcher matcher = MARKDOWN_HEADINGS_PATTERN_1.matcher(text); // Check for MarkDown headings #h1 - #h6
+		while (matcher.find()) {
+			int startPos = matcher.start() + node.getStartPosition();
+			int tokenIndex = tokenStartingAt(startPos);
+			Token headingToken = this.ctm.get(tokenIndex);
+			if (tokenIndex != 1) {
+				headingToken.putLineBreaksBefore(2);
+			}
+			int endPos = matcher.end() + node.getStartPosition();
+			int endIndex = tokenEndingAt(endPos - 1);
+			Token endingToken = this.ctm.get(endIndex);
+			endingToken.breakAfter();
+			headingToken.spaceBefore();
+		}
+
+		matcher = MARKDOWN_HEADINGS_PATTERN_2.matcher(text); // Check for MarkDown headings with styles '-- & ==='
+		while (matcher.find()) {
+			int startPos = matcher.start() + node.getStartPosition();
+			int tokenIndex = tokenStartingAt(startPos);
+			Token headingToken = this.ctm.get(tokenIndex);
+			if (tokenIndex != 1) {
+				headingToken.breakBefore();
+			}
+			headingToken.breakAfter();
+		}
+
+		List<ASTNode> fragments = node.fragments();
+		ArrayDeque<ASTNode> fragmentsDeque = new ArrayDeque<>(fragments);
+		handleMarkdownList(fragmentsDeque, this.markdownLeadingSpaces, 0, null);
+		handleMarkdownTable(fragments);
+
+		matcher = MARKDOWN_FENCES_PATTERN.matcher(text); // Check for Markdown snippet with styles '``` & ```'
+		while (matcher.find()) {
+			int startPos = matcher.start() + node.getStartPosition();
+			int tokenIndex = this.ctm.findIndex(startPos, ANY, true);
+			Token openingToken = this.ctm.get(tokenIndex);
+			int endPos = -1;
+			int tokenIndexLast = -1;
+			Token closingToken = null;
+			boolean shouldDisable = false;
+			String fenceCharsStart = matcher.group(1);
+		    String infoString = matcher.group(2);
+		    int fenceLengthStart = fenceCharsStart.length();
+			while (matcher.find()) {
+				shouldDisable = true;
+				endPos = matcher.start() + node.getStartPosition();
+				tokenIndexLast = this.ctm.findIndex(endPos, ANY, true);
+				closingToken = this.ctm.get(tokenIndexLast);
+				String fenceCharsEnd = matcher.group(1); // the fence itself
+			    int fenceLengthEnd = fenceCharsStart.length();
+				// Check if fences match
+				if (!fenceCharsStart.equals(fenceCharsEnd))
+					continue;
+
+				if (fenceLengthStart == fenceLengthEnd) {
+					if (tokenIndex > 1)
+						openingToken.breakBefore();
+					openingToken.breakAfter();
```

Comment:
there could be some info string tokens after openingToken, need to find the last one and break after that.

or maybe it would be simpler to use `disableFormatting` instead of `disableFormattingExclusively` so that openingToken and any whitespace/newlines also get disabled/preserved, then no need for breakAfter here.